File: APE1/Automovilestudiantesjava/Estudiante.java
package Automovilestudiantesjava;
import java.time.LocalDate;
import java.time.Period;

public class Estudiante{
        // Atributos
    private String cedula;
    private String nombres;
    private String apellidos;
    private LocalDate fechaNacimiento;

    // Vector para maximo notas
    private double[] notas;

        // Constructor
    public Estudiante(String cedula, String nombres, String apellidos,
                      LocalDate fechaNacimiento) {

        this.cedula = cedula;
        this.nombres = nombres;
        this.apellidos = apellidos;
        this.fechaNacimiento = fechaNacimiento;

        // Maximo 7 notas
        this.notas = new double[7];
    }

    // Getters
    public String getCedula() {
        return cedula;
    }

    public String getNombres() {
        return nombres;
    }

    public String getApellidos() {
        return apellidos;
    }

    public LocalDate getFechaNacimiento() {
        return fechaNacimiento;
    }

    public double[] getNotas() {
        return notas;
    }

    // Setters
    public void setCedula(String cedula) {
        this.cedula = cedula;
    }

    public void setNombres(String nombres) {
        this.nombres = nombres;
    }

    public void setApellidos(String apellidos) {
        this.apellidos = apellidos;
    }

    public void setFechaNacimiento(LocalDate fechaNacimiento) {
        this.fechaNacimiento = fechaNacimiento;
    }
    // Calcular edad
    public int getEdad() {
        return Period.between(fechaNacimiento, LocalDate.now()).getYears();
    }

    // Buscar una posición libre en el vector de notas
    public int buscarPosicionLibre() {

        for (int i = 0; i < notas.length; i++) {
            if (notas[i] == 0) {
                return i;
            }
        }

        return -1;
    }

    // Agregar nota
    public boolean agregarNota(double nota) {

        int posicion = buscarPosicionLibre();

        if (posicion == -1) {
            return false;
        }

        notas[posicion] = nota;
        return true;
    }

    // Modificar nota
    public boolean modificarNota(int posicion, double nuevaNota) {

        if (posicion < 0 || posicion >= notas.length) {
            return false;
        }

        if (notas[posicion] == 0) {
            return false;
        }

        notas[posicion] = nuevaNota;
        return true;
    }

    // Eliminar nota
    public boolean eliminarNota(int posicion) {

        if (posicion < 0 || posicion >= notas.length) {
            return false;
        }

        if (notas[posicion] == 0) {
            return false;
        }

        // Mover las notas siguientes hacia atras
        for (int i = posicion; i < notas.length - 1; i++) {
            notas[i] = notas[i + 1];
        }

        notas[notas.length - 1] = 0;

        return true;
    }

    // Cantidad de notas registradas
    public int cantidadNotas() {

        int cantidad = 0;

        for (int i = 0; i < notas.length; i++) {
            if (notas[i] != 0) {
                cantidad++;
            }
        }

        return cantidad;
    }

    // Calcular promedio del estudiante
    public double calcularPromedio() {

        double suma = 0;
        int cantidad = 0;

        for (int i = 0; i < notas.length; i++) {

            if (notas[i] != 0) {
                suma += notas[i];
                cantidad++;
            }
        }

        if (cantidad == 0) {
            return 0;
        }

        return suma / cantidad;
    }

    // Mostrar datos
    public void mostrarDatos() {

        System.out.println("Cedula: " + cedula);
        System.out.println("Nombres: " + nombres);
        System.out.println("Apellidos: " + apellidos);
        System.out.println("Fecha de nacimiento: " + fechaNacimiento);
        System.out.println("Edad: " + getEdad());
    }
}

File: APE1/Estudiante/Estudiante.cpp

#include "Estudiante.h"
#include <iostream>

Estudiante::Estudiante()
    : cedula(""), nombres(""), apellidos(""), fechaNacimiento(Fecha()) {
    for (int i = 0; i < 7; i++) notas[i] = 0;
}

Estudiante::Estudiante(const std::string& cedula, const std::string& nombres,
                       const std::string& apellidos, const Fecha& fechaNacimiento)
    : cedula(cedula), nombres(nombres), apellidos(apellidos),
      fechaNacimiento(fechaNacimiento) {
    // Máximo 7 notas (igual que "this.notas = new double[7];" en Java)
    for (int i = 0; i < 7; i++) notas[i] = 0;
}

std::string Estudiante::getCedula() const { return cedula; }
std::string Estudiante::getNombres() const { return nombres; }
std::string Estudiante::getApellidos() const { return apellidos; }
Fecha Estudiante::getFechaNacimiento() const { return fechaNacimiento; }
double* Estudiante::getNotas() { return notas; }

void Estudiante::setCedula(const std::string& c) { cedula = c; }
void Estudiante::setNombres(const std::string& n) { nombres = n; }
void Estudiante::setApellidos(const std::string& a) { apellidos = a; }
void Estudiante::setFechaNacimiento(const Fecha& f) { fechaNacimiento = f; }

int Estudiante::getEdad() const {
    return calcularEdad(fechaNacimiento);
}

int Estudiante::buscarPosicionLibre() const {
    for (int i = 0; i < 7; i++) {
        if (notas[i] == 0) {
            return i;
        }
    }
    return -1;
}

bool Estudiante::agregarNota(double nota) {
    int posicion = buscarPosicionLibre();
    if (posicion == -1) {
        return false;
    }
    notas[posicion] = nota;
    return true;
}

bool Estudiante::modificarNota(int posicion, double nuevaNota) {
    if (posicion < 0 || posicion >= 7) {
        return false;
    }
    if (notas[posicion] == 0) {
        return false;
    }
    notas[posicion] = nuevaNota;
    return true;
}

bool Estudiante::eliminarNota(int posicion) {
    if (posicion < 0 || posicion >= 7) {
        return false;
    }
    if (notas[posicion] == 0) {
        return false;
    }
    // Mover las notas siguientes hacia atrás
    for (int i = posicion; i < 7 - 1; i++) {
        notas[i] = notas[i + 1];
    }
    notas[7 - 1] = 0;
    return true;
}

int Estudiante::cantidadNotas() const {
    int cantidad = 0;
    for (int i = 0; i < 7; i++) {
        if (notas[i] != 0) {
            cantidad++;
        }
    }
    return cantidad;
}

double Estudiante::calcularPromedio() const {
    double suma = 0;
    int cantidad = 0;
    for (int i = 0; i < 7; i++) {
        if (notas[i] != 0) {
            suma += notas[i];
            cantidad++;
        }
    }
    if (cantidad == 0) {
        return 0;
    }
    return suma / cantidad;
}

void Estudiante::mostrarDatos() const {
    std::cout << "Cedula: " << cedula << std::endl;
    std::cout << "Nombres: " << nombres << std::endl;
    std::cout << "Apellidos: " << apellidos << std::endl;
    std::cout << "Fecha de nacimiento: " << fechaNacimiento << std::endl;
    std::cout << "Edad: " << getEdad() << std::endl;
}